modules/standard/DistributedBag.chpl

```diff
@@ -0,0 +1,1245 @@
+  other segment loses is proportional to how much it owns, ensuring a balance.
+*/
+config param WORK_STEALING_RATIO = 0.25;
+/*
+  The maximum amount of work to steal from a horizontal node's segment. This
+  should be set to a value, in megabytes, that determines the maximum amount of
+  data that should be sent in bulk at once. The maximum number of elements is
+  determined by: (WORK_STEALING_MAX_MEMORY_MB * 1024 * 1024) / sizeof(eltType).
+  For example, if we are storing 8-byte integers and have a 1MB limit, we would
+  have a maximum of 125,000 elements stolen at once.
+*/
+config param WORK_STEALING_MAX_MEMORY_MB : real = 1.0;
+/*
+  Determines whether or not we check if we are 'frozen'. These checks cause a
+  significant decline in performance, of about 33%. If the utility is not needed,
+  as it is more useful when reads are performed more often than writes, it may
+  be disabled for a raw performance boost. If the user attempts to freeze the
+  bag with this set to true, it merely returns false; all other checks are elided.
+*/
+config param BAG_NO_FREEZE = false;
+/*
+  The minimum number of elements a horizontal segment must have to become eligible
+  to be stolen from. This may be useful if some segments produce less elements than
+  others and should not be stolen from.
+*/
+config param WORK_STEALING_MINIMUM = 1;
+
+/*
+  A parallel-safe distributed multiset implementation that scales in terms of
+  both nodes and processors per node. This data structure is unordered and employs
+  its own work-stealing algorithm, and provides a means to obtain a privatized instance of
+  the data structure for maximized performance. This is an ideal backbone for a
+  means to distribute work across nodes.
+*/
+class DistributedBag : Collection {
+  pragma "no doc"
+  var targetLocDom : domain(1);
+  /*
+    The locales to allocate bags for and load balance across.
+  */
+  var targetLocales : [targetLocDom] locale;
+  pragma "no doc"
+  var pid : int = -1;
+
+  // Node-local fields
+  pragma "no doc"
+  var bag : Bag(eltType);
+  pragma "no doc"
+  var concurrentTasks : atomic int;
+  pragma "no doc"
+  var frozenState : atomic int;
+
+  proc DistributedBag(type eltType, targetLocales : [?targetLocDom] locale = Locales) {
+    bag = new Bag(eltType, this);
+    this.targetLocDom = targetLocDom;
+    this.targetLocales = targetLocales;
+    pid = _newPrivatizedClass(this);
+  }
+
+  pragma "no doc"
+  proc DistributedBag(other, pid, type eltType = other.eltType) {
+    bag = new Bag(eltType, this);
+    this.targetLocDom = other.targetLocDom;
+    this.targetLocales = other.targetLocales;
+    this.pid = pid;
+  }
+
+  pragma "no doc"
+  proc dsiPrivatize(pid) {
+    return new DistributedBag(this, pid);
+  }
+
+  pragma "no doc"
+  proc dsiGetPrivatizeData() {
+    return pid;
+  }
+
+  pragma "no doc"
+  inline proc getPrivatizedThis {
+    if this.locale == here {
+      return this;
+    }

```

Comment:
@mppf 

Is [this](https://github.com/chapel-lang/chapel/blob/86164f1a7433a12855cd5a002522a4639194c1e4/modules/internal/ChapelArray.chpl#L213-L216) comment still relevant? If so, then we need to avoid excessive calls to get a privatized instance if its managed by Locale 0. If not, I can easily remove it.